tests/utils/tensorflow/test_rasa_layers.py

```diff
@@ -0,0 +1,855 @@
+            {
+                "feature_type": "arbitrary",
+                "feature_type_signature": [feature_signature_sparse_1],
+            },
+            units_sparse_to_dense,
+        ),
+        # ConcatenateSparseDense layer with only dense features
+        (
+            ConcatenateSparseDenseFeatures,
+            model_config_basic,
+            {
+                "feature_type": "arbitrary",
+                "feature_type_signature": [feature_signature_dense_1],
+            },
+            units_1,
+        ),
+        # FeatureCombining layer with sequence- and sentence-level features, doing
+        # dimension unifying
+        (
+            RasaFeatureCombiningLayer,
+            model_config_basic,
+            {"attribute_signature": attribute_signature_basic},
+            units_concat,
+        ),
+        # FeatureCombining layer with sequence- and sentence-level features, no
+        # dimension unifying
+        (
+            RasaFeatureCombiningLayer,
+            model_config_basic,
+            {
+                "attribute_signature": {
+                    SEQUENCE: [feature_signature_dense_1],
+                    SENTENCE: [feature_signature_dense_1],
+                }
+            },
+            units_1,
+        ),
+        # FeatureCombining layer with sentence-level features only
+        (
+            RasaFeatureCombiningLayer,
+            model_config_basic,
+            {
+                "attribute_signature": {
+                    "sequence": [],
+                    "sentence": [feature_signature_dense_1],
+                },
+            },
+            units_1,
+        ),
+        # FeatureCombining layer with sequence-level features only
+        (
+            RasaFeatureCombiningLayer,
+            model_config_basic,
+            {
+                "attribute_signature": {
+                    "sequence": [feature_signature_dense_1],
+                    "sentence": [],
+                },
+            },
+            units_1,
+        ),
+        # Sequence layer with mixed features, hidden layers and transformer
+        (
+            RasaSequenceLayer,
+            model_config_transformer,
+            {"attribute_signature": attribute_signature_basic},
+            units_transformer,
+        ),
+        # Sequence layer with mixed features, hidden layers, no transformer
+        (
+            RasaSequenceLayer,
+            model_config_basic,
+            {"attribute_signature": attribute_signature_basic},
+            units_hidden_layer,
+        ),
+        # Sequence layer with mixed features, no hidden layers, no transformer
+        (
+            RasaSequenceLayer,
+            model_config_basic_no_hidden_layers,
+            {"attribute_signature": attribute_signature_basic},
+            units_concat,
+        ),
+    ],
+)
+def test_layer_gives_correct_output_units(
+    layer_class: Callable[..., Any],  # TODO change the callable signature? (also below)
```

Comment:
Here I pass class constructors as arguments, and their input signatures differ since they are constructors of the three different classes. Shall I keep the `Callable`'s signature vague like this and not bother with specifying any details? 🤔

Reply:
The right annotation would probably be `Type[tf.keras.layers.Layer]`
see https://docs.python.org/3/library/typing.html#typing.Type

Reply:
Ah, right, thanks! How about I make it a bit more explicit: `Union[Type[ConcatenateSparseDenseFeatures, RasaFeatureCombiningLayer, RasaSequenceLayer]]`

Reply:
I think it would have to be `Union[Type[Concate...], Type[RasaFeature..], Type[RasaSeq..]]`, but not entirely sure. In general I'd be in favor of `Type[Layer]` and import that class from keras. Also fits neatly with the variable name `layer_class`. If you just take the parent class you won't need to change the type annotation when you add another layer.

Reply:
Right, I think I'll go with the parent class, thanks 🙂